### ISSUE TITLE
Issue with duk_instanceof

### ISSUE THREAD

**Issue by edoardomazzella** [svaarala/duktape#1604]
Hello!

duk_get_global_string(ctx,"IOBUS");
duk_push_object(ctx);

if(!duk_instanceof(ctx,-1,-2)) ...

This is not the real code, so i know that it has no sense, but i found this problem: when duk_instanceof is called, even in this case, with an object as first argument it always throws an error "object required, found undefined". Are you sure it works fine?
Thank you!

**Comment by svaarala** [comment]
I'm never completely sure about anything but at least the API tests works: https://github.com/svaarala/duktape/blob/master/tests/api/test-instanceof.c.

If you provide a reproducible example I can test it.

**Comment by svaarala** [comment]
At least the API test seems to lack testing for negative indices. They go through helpers that should deal with them correctly, but I'll check if that makes a difference for some reason. The internal call sites to `duk_instanceof()` only use non-negative indices so a bug there wouldn't be caught.

**Comment by svaarala** [comment]
Could you try this?

```
        duk_eval_string(ctx, "new RangeError('aiee')");
        duk_eval_string(ctx, "Error");

        printf("top=%ld\n", (long) duk_get_top(ctx));
        printf("%ld\n", (long) duk_instanceof(ctx, -2, -1));
```

This print '1' for me, i.e. `new RangeError('aiee') instanceof Error` is true.

**Comment by edoardomazzella** [comment]
yes it prints 1...
My code is something like this:

static duk_ret_t IOBUS_dtor(duk_context *ctx){

  return 0;
}

static duk_ret_t IOBUS_ctor(duk_context *ctx){

  if (!duk_is_constructor_call(ctx))
    return duk_type_error(ctx, "Function IOBUS_ctor must be called as"
        " constructor");

  duk_push_this(ctx);

  duk_dup(ctx,0);
  duk_put_prop_string(ctx,-2,"mask");

  duk_dup(ctx,1);
  duk_put_prop_string(ctx,-2,"offset");

  // Store the function destructor
  duk_push_c_function(ctx, IOBUS_dtor, 0);
  duk_set_finalizer(ctx, -2);

  return 0;
}

void scriptly_create_IOBUS_prototype(duk_context *ctx){
  duk_push_c_function(ctx, IOBUS_ctor, 2);
  duk_put_global_string(ctx, "IOBUS");
}

static duk_ret_t duk_read_bus(duk_context *ctx){

  IOBus bus;

  duk_get_global_string(ctx,"IOBUS");

  printf("%d",duk_instanceof(ctx,0,-1)); //here i get an error

// ... ... ...

return 0;
}


Ecmascript side:
"var f = new IOBUS(6,7);\n"
"readBus(f);"

**Comment by svaarala** [comment]
I think the issue might be that `IOBUS`, while a constructor, does not have a `.prototype` property which instanceof expects. Unlike Ecmascript functions, Duktape/C functions don't have a `.prototype` by default.

The Ecmascript equivalent of the issue would be:
```
duk> function Constructor() {}
= undefined
duk> Constructor.prototype = void 0;
= undefined
duk> print({} instanceof Constructor)
TypeError: object required, found undefined (stack index -1)
    at [anon] (duk_api_stack.c:1633) internal
    at global (input:1) preventsyield
```

Without a `.prototype` property on the constructor instance objects inherit directly from Object.prototype. You can add an explicit `.prototype` which should allow instanceof to work as expected.

**Comment by svaarala** [comment]
So to be clear, the "found undefined" refers to the `.prototype` being undefined. The error message could be improved to e.g.:

```
function has no .prototype property
```

**Comment by edoardomazzella** [comment]
thanks

**Comment by svaarala** [comment]
Closing, error message improvement is implemented by #1725.
